src/R/Components/Impl/Containers/Implementation/ContainerManager.cs

```diff
@@ -0,0 +1,143 @@
+    internal class ContainerManager : IContainerManager {
+        private readonly IContainerService _containerService;
+        private readonly IFileSystem _fileSystem;
+        private readonly CountdownDisposable _containersChangedCountdown;
+        private ImmutableArray<IContainer> _containers;
+        private ImmutableArray<IContainer> _runningContainers;
+        private CancellationTokenSource _updateContainersCts;
+        private event Action ContainersChanged;
+
+        public ContainerManager(IRInteractiveWorkflow interactiveWorkflow) {
+            _containerService = interactiveWorkflow.Services.GetService<IContainerService>();
+            _fileSystem = interactiveWorkflow.Services.FileSystem();
+            _updateContainersCts = new CancellationTokenSource();
+            _containers = ImmutableArray<IContainer>.Empty;
+            _runningContainers = ImmutableArray<IContainer>.Empty;
+            _containersChangedCountdown = new CountdownDisposable(StartUpdatingContainers, EndUpdatingContainers);
+        }
+
+        public IReadOnlyList<IContainer> GetContainers() => _containers;
+        public IReadOnlyList<IContainer> GetRunningContainers() => _runningContainers;
+        public IDisposable SubscribeOnChanges(Action containersChanged) {
+            ContainersChanged += containersChanged;
+            _containersChangedCountdown.Increment();
+            return Disposable.Create(() => {
+                ContainersChanged -= containersChanged;
+                _containersChangedCountdown.Decrement();
+            });
+        }
+
+        public async Task StartAsync(string containerId, CancellationToken cancellationToken) {
+            try {
+                await _containerService.StartContainerAsync(containerId, cancellationToken);
+            } finally {
+                await UpdateContainersOnceAsync(cancellationToken);
+            }
+        }
+
+        public async Task StopAsync(string containerId, CancellationToken cancellationToken) {
+            try {
+                await _containerService.StopContainerAsync(containerId, cancellationToken);
+            } finally {
+                await UpdateContainersOnceAsync(cancellationToken);
+            }
+        }
+
+        public async Task DeleteAsync(string containerId, CancellationToken cancellationToken) {
+            try {
+                await _containerService.DeleteContainerAsync(containerId, cancellationToken);
+            } finally {
+                await UpdateContainersOnceAsync(cancellationToken);
+            }
+        }
+
+        public async Task<IContainer> CreateLocalDockerAsync(string name, string username, string password, string version, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrWhiteSpace(version)) {
+                version = "3.4.1";
+            }
+
+            var dockerImageContent = $@"FROM kvnadig/rtvsd-ub1604:{version}
```

Comment:
I'd put this in a file

Reply:
Ok, should I just put it in the folder with RTVS assemblies?

Reply:
Yes, it will do. We (or user) can then edit it without touching the code.

Reply:
https://github.com/Microsoft/RTVS/issues/3947 - for some reason `CopyToOutputDirectory` doesn't work in this project (works perfectly in `Microsoft.UnitTests.Core.Linux`).